### ISSUE TITLE
[Android] Allow setting the 24 hour format instead follow user locale

### ISSUE THREAD

**Issue by richardtks** [henninghall/react-native-date-picker#130]
I am really grateful this date picker has the consistent UI and behavior across the ios and android platform and this is what I need for my project. Really appreciate the effort and hard work of doing this. 😄 

Understand that for the android platform, the 24 format is following the user locale setting set in the phone. Is it possible to override user locale setting by adding a prop called `is24Format`?

**Comment by henninghall** [comment]
Hi, glad you like it! I have got this question before but I don't really understand the use case of it. Shouldn't we let the user decide how they want to see their dates? 🤔

**Comment by richardtks** [comment]
@henninghall, thank you for the fast response. 😄  I understand your point, for my case, it just simply request from our client to use the 24 hours regardless of user locale. 

However, there is one possible use case I can think of is that if the app itself allow the user to set the time format for that particular app without affecting the system locale setting, with `is24Format` props we can actually achieve that. Is this make sense to you?

**Comment by henninghall** [comment]
Yeah I guess that would be a valid reason, if you cannot convince a client that it is a bad idea 😉Just remembered the biggest problem with this, the UIDatePicker from IOS does not support setting the 24h value from anything else than the locale. And I would not like to add an android specific prop for this, it would be confusing. 

What I might consider is adding a prop for letting the locale decide if to use 24h mode even on android (`useLocaleToDecide24format`). Would that help you in your situation?

**Comment by richardtks** [comment]
@henninghall, thank you for your understanding. Yeah, I aware that UIDatePicker from IOS does not support setting the 24h, for now, I would just `fr` locale which use 24h. 

`useLocaleToDecide24format` only change the 24h based on locale? Wouldn't it better if user set the `locale` props, then we use exactly the prop being set for android? (p/s: ios already support this prop)

**Comment by henninghall** [comment]
Yeah that was how it actually worked before on Android. I still think it is a better to listen to the users 24h preference in settings. I just saw that iOS app "Reminders" actually changes the picker 24h preference depending on that setting, not on the locale 😮So it seems to be possible in some way. I that case we could add a "override-24-h-setting" prop that works on both platforms. I will look into this after christmas, I have a lot to do at the moment.

**Comment by FedericoNavarro** [comment]
Hello! I also need this functionality, since in my design, I have two of this Time Pickers at one each other side, and on 12hs it is to big. I also tried by reducing the font size, but it doesn't seems to work. Any idea? Thanks!

**Comment by olegmilan** [comment]
@FedericoNavarro you can try this solution as a temp workaround
https://github.com/henninghall/react-native-date-picker/issues/111#issuecomment-562577758

**Comment by grifotv** [comment]
I believe a prop `is24Hour` would be extremely helpful even if it's for Android only, like this one:
- https://github.com/react-native-community/react-native-datetimepicker#is24hour-optional-android-only

For instance, my use case would be a parking app, where users have the ability to set either:
1. End parking date and time, `mode="datetime"`
2. Or duration of parking, `mode="time"` // but this would only make sense if it's in the 24h format (start at 00:00) as the 12h format doesn't start at zero (start at 12:00 AM).

With the currently implementation, I'm forced to do this workaround https://github.com/henninghall/react-native-date-picker/issues/130#issuecomment-577340060

**Comment by tboerc** [comment]
I believe that the Android only prop for 24 hours would be very good too, cause in my app, I use the picker to get the amount of time that some task will have, and if it uses AM/PM, is very confusing, so 24 hours would be better.

**Comment by andreialecu** [comment]
@henninghall I notice you're back in action here. Would be awesome if this issue could be addressed as part of the current set of changes. Seems like it would be easy enough to support based on: https://github.com/henninghall/react-native-date-picker/issues/111#issuecomment-562577758

**Comment by henninghall** [comment]
@andreialecu I am definitely considering it but I am not totally convinced that it is the best approach. 

1. In some situations I believe that a duration mode would be a better solution, for instance https://github.com/henninghall/react-native-date-picker/issues/130#issuecomment-578065404. Im tracking duration/countdown mode in this issue https://github.com/henninghall/react-native-date-picker/issues/199
2. In other situations I think it might lead to layout problems on iOS if the developer assumes that am/pm won't be shown. (not possible to disable on iOS) and it would probably be better to take in account that am/pm might be shown. For instance https://github.com/henninghall/react-native-date-picker/issues/130#issuecomment-566147985. 
3. It doesn't feel very user-friendly to hide am/pm although the user has enabled that in the settings.

I'm still open to change my mind with a good argument though! 😉 
What is your use case?

**Comment by andreialecu** [comment]
I'm not planning to use it for duration fwiw.

Maybe `prefer24h={true}` or `androidPrefer24h={true}`? In my case the app allows changing the locale from within its settings. 

The custom locale generally uses 24 hour formats and it can confuse users if they see am/pm. It's not a huge deal, but seeing that there's a flag for it in the underlaying library I don't see why it wouldn't be exposed. :)

**Comment by henninghall** [comment]
Okay that's cool, would you mind explaining what you mean by "custom locale"? 

As I see it, the reason for not exposing it would be the risk of developers mistakenly creating a bad user experience. That might also be the reason for Apple not wanting to expose this setting in the iOS datepicker. 

Would a prop called something like "use24hourPreferenceFromLocale" solve the problem for you? If enabled the 24hour  preference would be determined by the locale on both platforms (as it already is on iOS).

**Comment by andreialecu** [comment]
In practice it may be common for the the user's device to be set to an English locale, by default. For example, Romanian uses 24h time - as probably most European countries. 

Based on my analytics about half of users here set their OS-level phone language to Romanian and the other half use English. Even though the phone may be in English, most romanian speaking users would pick Romanian for specific apps if presented with the choice. I would like to enforce 24h time if the user picked Romanian as I'm pretty sure 99% of them would prefer that format.

Additionally, I'm also working on a Kiosk-like app for Android devices. A similar example is ATMs which let you choose your language from the Welcome screen.

Since the app is going to be placed on an Android tablet on a stand somewhere in a crowded place with people from all over the world, being able to change the 24h format on the fly would be great.

**Comment by henninghall** [comment]
Interesting, I think I understand! So if it worked like the iOS picker (24h-format is determined by selected locale, Romanian locale would always show 24h) it would work for your use-case, right?

**Comment by andreialecu** [comment]
Yes. I think so. Having it follow the locale settings is what I need.

**Comment by henninghall** [comment]
Then I'll start adding that feature, and add a "force24h" prop later on if someone finds a situation were it is really no way around it.
